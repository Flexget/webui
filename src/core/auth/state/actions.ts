import { asyncAction, ActionsUnion, RequestsOfType as ROT } from 'core/request/state/util';

import { StatusError } from 'utils/fetch';

export const enum Constants {
  LOGIN = '@flexget/auth/LOGIN',
  LOGOUT = '@flexget/auth/LOGOUT',
}

const asyncActions = {
  login: {
    request: (username: string, password: string) =>
      asyncAction.request(Constants.LOGIN, { username, password }),
    success: () => asyncAction.success(Constants.LOGIN),
    failure: (err: StatusError) => asyncAction.failure(Constants.LOGIN, err),
  },
  logout: {
    request: () => asyncAction.request(Constants.LOGOUT),
    success: () => asyncAction.success(Constants.LOGOUT),
    failure: (err: StatusError) => asyncAction.failure(Constants.LOGOUT, err),
  },
};

export type ActionTypes = ActionsUnion<typeof asyncActions>;
export type RequestsOfType<T extends Constants> = ROT<ActionTypes, T>;

export default asyncActions;
