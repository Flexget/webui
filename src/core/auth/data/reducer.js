import { LOGIN, LOGOUT } from 'core/auth/data/actions';
import { ERROR_STATUS } from 'core/status/data/actions';
import { GET_VERSION } from 'core/version/data/actions';

export default function reducer(state = {}, action) {
  switch (action.type) {
    case LOGIN:
    case GET_VERSION:
      return {
        loggedIn: true,
      };
    case LOGOUT:
      return {};
    case ERROR_STATUS:
      if (action.payload.statusCode === 401) {
        return {};
      }
    default: // eslint-disable-line no-fallthrough
      return state;
  }
}
