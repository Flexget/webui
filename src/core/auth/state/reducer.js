import { ERROR_STATUS } from 'core/status/state/actions';
import { GET_VERSION } from 'core/version/state/actions';
import { LOGIN, LOGOUT } from './actions';

export default function reducer(state = {}, action) {
  switch (action.type) {
    case LOGIN:
    case GET_VERSION:
      return {
        loggedIn: true,
      };
    case LOGOUT:
      return {};
    case ERROR_STATUS:
      if (action.payload.statusCode === 401) {
        return {};
      }

    // falls through
    default:
      return state;
  }
}
