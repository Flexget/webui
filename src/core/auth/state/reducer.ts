import {
  Constants as StatusConstants,
  ActionTypes as StatusActionTypes,
} from 'core/status/state/actions';
import {
  Constants as VersionConstants,
  ActionTypes as VersionActionTypes,
} from 'core/version/state/actions';
import { Constants, ActionTypes } from './actions';

export interface State {
  loggedIn?: Boolean;
}

const initState: State = {
  loggedIn: false,
};

export default function reducer(
  state = initState,
  action: ActionTypes | StatusActionTypes | VersionActionTypes,
): State {
  switch (action.type) {
    case Constants.LOGIN:
    case VersionConstants.GET_VERSION:
      return {
        loggedIn: true,
      };
    case Constants.LOGOUT:
      return {};
    case StatusConstants.ERROR_STATUS:
      if (action.payload.statusCode === 401) {
        return {};
      }
    // falls through
    default:
      return state;
  }
}
