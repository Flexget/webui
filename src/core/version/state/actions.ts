import statusActions from 'core/status/state/actions';
import { RequestsOfType as ROT } from 'core/status/state/types';
import { action, ActionsUnion } from 'utils/actions';
import { StatusError } from 'utils/fetch';

export const enum Constants {
  GET_VERSION = '@flexget/version/GET_VERSION',
}

export interface Version {
  apiVersion: string;
  flexgetVersion: string;
  latestVersion: string;
}

const actions = {
  getVersion: {
    request: () => statusActions.load(Constants.GET_VERSION),
    success: (data: Version) => action(Constants.GET_VERSION, data),
    failure: (err: StatusError) => statusActions.error(Constants.GET_VERSION, err),
  },
};

export type ActionTypes = ActionsUnion<typeof actions>;
export type RequestsOfType<T extends Constants> = ROT<ActionTypes, T>;

export default actions;
