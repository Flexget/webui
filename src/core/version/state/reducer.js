import { GET_VERSION } from './actions';

export default function reducer(state = {}, action) {
  switch (action.type) {
    case GET_VERSION: {
      const { apiVersion: api, flexgetVersion: flexget, latestVersion: latest } = action.payload;

      return { api, flexget, latest };
    }
    default:
      return state;
  }
}
