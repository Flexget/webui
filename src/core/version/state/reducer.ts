import { createSuccessCheck } from 'core/request/state/util';
import { Constants, ActionTypes } from './actions';

export interface State {
  api?: string;
  flexget?: string;
  latest?: string;
}

const initState: State = {};

export default function reducer(state = initState, action: ActionTypes): State {
  const check = createSuccessCheck(action);
  switch (true) {
    case check(Constants.GET_VERSION): {
      const { apiVersion: api, flexgetVersion: flexget, latestVersion: latest } = action.payload;

      return { api, flexget, latest };
    }
    default:
      return state;
  }
}
