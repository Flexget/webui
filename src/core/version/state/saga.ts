import { call, put, takeLatest } from 'redux-saga/effects';
import { delay } from 'redux-saga';
import { requesting } from 'core/status/state/util';
import { get } from 'utils/fetch';
import actions, { Constants } from './actions';

export function* getVersion() {
  try {
    yield call(delay, 500);
    const { data } = yield call(get, '/server/version');
    yield put(actions.getVersion.success(data));
  } catch (err) {
    yield put(actions.getVersion.failure(err));
  }
}

export default function* saga() {
  yield takeLatest(requesting(Constants.GET_VERSION), getVersion);
}
