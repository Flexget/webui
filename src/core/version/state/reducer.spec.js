import reducer from 'core/version/state/reducer';
import { GET_VERSION } from 'core/version/state/actions';

describe('core/version/data/reducer', () => {
  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should get version on GET_VERSION', () => {
    expect(
      reducer(undefined, {
        type: GET_VERSION,
        payload: {
          apiVersion: '1.1.2',
          flexgetVersion: '2.10.11',
          latestVersion: '2.10.60',
        },
      }),
    ).toMatchSnapshot();
  });
});
