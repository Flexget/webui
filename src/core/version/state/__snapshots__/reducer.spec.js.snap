// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`core/version/data/reducer should get version on GET_VERSION 1`] = `
Object {
  "api": "1.1.2",
  "flexget": "2.10.11",
  "latest": "2.10.60",
}
`;

exports[`core/version/data/reducer should return the initial state 1`] = `Object {}`;
